### PR TITLE
Free photo library: add free photo library using Pexels data source

### PR DESCRIPTION
**Pull request by notnownikki**
This adds a new external photo library, using the `pexels` external media source.

When the `external-free-photo-library` config is set to `true`, the new source should appear in the media library top corner menu, and the tinymce "Add" menu:

1 - new source in icon dropdown

2 -new item in tinymce add menu

By default, no images are shown, and you can search for photos:

3 - free photo search

You can copy multiple items to your WP library:

4 - find cute kittens

5 - copy to media library

Or you can insert a single image into a blog post:

6 - insert single image

7 - image in blog post

## Testing

* Test the above steps on a wpcom site
* Test the above steps on a Jetpack site
* Check that you can copy from the free photos using the media library
* ~~Check that you can set the site icon from the free photo library.~~ Check the free photo library is not available when you set the site icon, as it provides images too big for the icon editor to handle.

### CODE REVIEW

File: client/my-sites/media-library/external-media-header.jsx [hunk 1]
```diff
@@ -111,16 +111,30 @@ class MediaLibraryExternalHeader extends React.Component {
...
 
 		return (
 			<Card className="media-library__header">
-				<Button compact disabled={ this.state.fetching } onClick={ this.handleClick }>
-					<Gridicon icon="refresh" size={ 24 } />
+				{ 'pexels' === source && this.renderPexelsAttribution() }
+
+				{ 'pexels' !== source && (
+					<Button compact disabled={ this.state.fetching } onClick={ this.handleClick }>
+						<Gridicon icon="refresh" size={ 24 } />
```
Comment (notnownikki) [inline client/my-sites/media-library/external-media-header.jsx]: Makes no sense to have the refresh button on the free photo search.

File: client/my-sites/media-library/list-no-content.jsx [hunk 2]
```diff
@@ -76,7 +82,7 @@ class MediaLibraryListNoContent extends Component {
 					{ this.props.translate( 'Upload Media' ) }
 				</UploadButton>
 			);
-		} else if ( this.props.source ) {
+		} else if ( 'google_photos' === this.props.source ) {
 			line = this.props.translate( 'New photos may take a few minutes to appear.' );
 		}
```
Comment (notnownikki) [inline client/my-sites/media-library/list-no-content.jsx]: this only applies to Google Photos now

File: client/lib/media/actions.js [hunk 3]
```diff
@@ -94,6 +94,12 @@ MediaActions.fetchNextPage = function( siteId ) {
 	if ( ! query.source ) {
 		wpcom.site( siteId ).mediaList( query, mediaReceived );
 	} else {
+		if ( 'pexels' === query.source ) {
+			// Pexels limits the number of API calls per month, and if we had the default
+			// limit of 20 media items, it takes at least 2 API calls to fill the first page
+			// with results. This saves us some of those API calls.
+			query.number = 40;
```
Comment (johngodley) [inline client/lib/media/actions.js]: I'm not keen on this special hardcoded limit, especially in such a low level position. Most APIs have limits - Google Photos does too - and Pexels have given us a very generous amount that I don't think we'll run out of soon.

If we do start having limit problems, or the WPCOM API gets abused which in turn causes rate limiting, then it should be tackled at a higher level in a consistent manner that applies across all services, with values stored in a central location

Comment (notnownikki) [reply client/lib/media/actions.js]: That seems sensible, I'll get this removed.

File: client/my-sites/media-library/content.scss [hunk 4]
```diff
@@ -29,3 +29,8 @@
 	margin: 0 auto;
 	text-align: center;
 }
+
+.media-library__pexels-attribution {
+	display: inline-block;
+	margin-right: 1em;
+}
```
[the remark this answers is not in the document]
Comment (notnownikki) [reply client/my-sites/media-library/content.scss]: Fixed

**Comment by notnownikki** [comment]
The two fixes we've agreed to try here:

1) Disable the free photo library source when selecting a site icon.

This will be done by telling the media modal to disable large image sources, and that will pass down to the media library (and therefore, the list of data sources) which sources to disable.

2) Reset the query on insert / copy / cancel.

This will stop existing queries getting messed with in the cupcake scenario.

**Comment by notnownikki** [comment]
Fixing up from @johngodley's review...

* Disabled `pexels` source in the site icon chooser
* Test added to make sure that disabling sources in the MediaLibraryDataSource works
* Config feature flag names, so we have 'external-media/*'
* When inserting a single image from an external source, reset the search and clear the query so that the "cupcake glitch" is hidden. There doesn't seem to be anything we can do to fix the glitch, it's a byproduct of how the pointers and transient images work, but this fix does hide it well enough, and doesn't impact any other code. Other fixes would have us modifying and/or adding special cases to CREATE_MEDIA_ITEM, and that's a whole bunch of nope because of how many different user actions get routed through that.

**Comment by notnownikki** [comment]
Ugh, that component test...

First, that DOM warning is exposing an existing problem. We have buttons within buttons. Meh.

So, enzyme+react16 is troublesome. In the test I wanted to click the button, and make sure the menu items appear. That relies on the child components updating, because we generate the menu items and the child components render them as `{ this.props.children }`.

Unfortunately, clicking the button sets the state on the component, re-renders the children, and _then_ passes them updated props, so the children never appear because there's logic there to not render them if the props say they're not visible. Therefore, we need to mock, and use a component that renders `{ this.props.children }` no matter what. I'll try the `const ShowChildren = props => <div>{ props.children }</div>; suggestion, but I couldn't get it working when I tried that before, there were no props coming through. But I'll try again.

Config is mocked in lots of tests, usually when we want to test things with all features enabled (most tests return `true` for all calls to `isEnabled`) but if we do that here, then we also enable `electron` which causes a build failure, so we only enable the external media options.

So today, I'll split off the `hasRefresh` and `hasAttribution` into a separate PR, as it's not directly tied to sorting out pexels and the site icon issue, and then this should be good to go when I get back from my AFK time, I hope!

**Comment by jasmussen** [comment]
Just tested this. It works well, works as intended, and John's (AFK) feedback above has been addressed. 

Given this is being launched for staging, horizon and wpcalypso, and code checks out, let's merge this early in the week so we can see that it works as intended.

Nice work! Ship it! 👍 👍

**Pull request merged**